### ISSUE TITLE
ISARIC

### ISSUE THREAD

**Issue by sebbacon** [opensafely-core/cohort-extractor#614]
Next steps: TPP to import the data into the OpenSAFELY database.

Currently (morning of 14 October) blocked on conversations / clarifications about linkage, between Chris and the ISARIC team

**Comment by alexwalkerepi** [comment]
There are two tables:

`ISARIC_Patient_Data_TopLine`
 - one row per patient
 - contains a mix of:
   - date columns
   - categorical strings, for example a boolean `"Yes", "No", "(Missing)"` or a unit description such as `"mmol/L", "mg/dL", "(Missing)"`
   - numerical columns

`ISARIC_Patient_Data`
 - multiple rows per patient (~8 per patient on average)
 - contains a similar mix of column types as the topline table, but the column names are not the same.

Both tables are _very_ wide, so I don't think it's practical to write specialised functions to return each specific field. Rather I think that  we should write functions to return the contents of user specified columns. I understand that the data have been pre-processed such that returning the raw values in a column would be fine. There may be a need for date filtering in some cases.

## ISARIC_Patient_Data_TopLine
Being one row per patient already, this feels straightforward, and I would imagine a study definition code would look something like:
### Date variable
```py
    isaric_assessment_date=patients.with_an_isaric_record(
        return_column="symp_or_assess_date",
        return_expectations={"incidence": 0.1, "date": {"earliest": "2020-02-01"}}
    ),
```

### Categorical variable:
```py
    isaric_fever=patients.with_an_isaric_record(
        between=["2020-02-01", "2020-06-01"],
        date_filter_column="dsstdat",
        return_column="fever",
        return_expectations={
            "rate": "universal",
            "category": {
                "ratios": {
                    "Yes": 0.7,
                    "No": 0.1,
                    "(Missing)": 0.2,
                }
            },
        },
    ),
```

### Float variable:
```py
    isaric_sysbp=patients.with_an_isaric_record(
        between=["2020-02-01", "2020-06-01"],
        date_filter_column="dsstdat",
        return_column="sysbp_vsorres",
        return_expectations={
            "incidence": 0.8,
            "float": {"distribution": "normal", "mean": 21, "stddev": 5},
        },
    ),
```

## ISARIC_Patient_Data
This table is multiple rows per patient and many columns are very sparsely populated. I think the pseudocode would likely look very similar to above, however:

The difficulty is how to reduce the data to one row per patient. Date is one possibility using e.g. `find_first_match_in_period=True`, but I need to speak to ISARIC users to understand whether you'd always expect certain date fields to be present when querying different fields.

**Comment by sebbacon** [comment]
So from memory all the current strategies to reduce are date-based:

* first or last by date within date range
* first or last by datetime within day
* mean value within date range
* mean value within day

What kinds of fields are you imaginging might have no date, out of interest?

**Comment by alexwalkerepi** [comment]
> What kinds of fields are you imaginging might have no date, out of interest?

Depending on the date field, dates only exist in ~25% or fewer of the rows. My assumption is that:
- if a researcher is interested in a specific field, there's _probably_ a complete date that relates to that field.
- The wideness of the data and the multiple date fields combine to create the date missingness.

I can dig further into it with some queries, but it might be better to talk to someone from ISARIC. If my assumption is correct they might be able to point to which date fields relate to which `str`/`int` fields.

**Comment by alexwalkerepi** [comment]
I spoke to @JamesAFarrell from ISARIC about this, two main conclusions from this are:

1. The one row per patient `ISARIC_Patient_Data_TopLine` table is a transformed version of the multiple rows per patient `ISARIC_Patient_Data`. The topline table should be sufficient for a large number of use cases, therefore that would be where to start to get things moving quickly.
2. We can likely think about the `ISARIC_Patient_Data` table later. It would only be required if you were looking at a sequence of events (e.g. CRP tests), and wanted to look at the trajectory. I explained that the current OpenSAFELY data model is focused on getting one row per patient data, but there are workarounds to get sequential events if needed. He confirmed my assumption above that if you were interested in e.g. a CRP test, then there should always be a corresponding date associated with that test, but which date field it was would depend on the field you were querying.

**Comment by sebbacon** [comment]
Next step: @iaindillingham to review / turn into a specification for the work / request clarifications as necessary

**Comment by iaindillingham** [comment]
@amirmehrkar would also like a warning added to the documentation about the use of ISARIC. For example wording, see opensafely/documentation#482.

**Comment by iaindillingham** [comment]
Closing, as this was implemented by #730
